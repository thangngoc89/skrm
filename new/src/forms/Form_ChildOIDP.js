import React from "react";
import { Formik, Form, Field, FastField } from "formik";
import { Box as GBox, Heading, Text, Button } from "grommet";
import {
  RadioGroup,
  SelectGroup,
  TextInput,
  DottedLabel,
  Box,
} from "../components";
import MountPortal from "../MountPortal";
import FormikAutosave from "../FormikAutosave";
import { hoatDong, lietkeOptions } from "./Form_ChildOIDP_data";
import {
  validate as handleValidation,
  REQUIRED,
} from "./Form_ChildOIDP_validate";
import * as notify from "../Notify";

export const blankInitialValues = {
  coKhoChiu: null,
  lietke: [],
  lietkeCustom: "",
  "1-mucdo": null,
  "1-tansuat": null,
  "1-nguyennhan": [],
  "2-mucdo": null,
  "2-tansuat": null,
  "2-nguyennhan": [],
  "3-mucdo": null,
  "3-tansuat": null,
  "3-nguyennhan": [],
  "4-mucdo": null,
  "4-tansuat": null,
  "4-nguyennhan": [],
  "5-mucdo": null,
  "5-tansuat": null,
  "5-nguyennhan": [],
  "6-mucdo": null,
  "6-tansuat": null,
  "6-nguyennhan": [],
  "7-mucdo": null,
  "7-tansuat": null,
  "7-nguyennhan": [],
  "8-mucdo": null,
  "8-tansuat": null,
  "8-nguyennhan": [],
};

const lietkeName = "lietke";
const lietkeCustomName = "lietkeCustom";

const Title = ({ title, subtitle }) => {
  return (
    <Box margin={{ bottom: "small" }}>
      <Heading level="2" color="brand">
        {title}
      </Heading>
      <Text size="medium" level={2}>
        {subtitle}
      </Text>
    </Box>
  );
};

const Section = ({ children, ...props }) => {
  return (
    <GBox as="section" margin={{ bottom: "large" }} {...props}>
      {children}
    </GBox>
  );
};
const Part2 = ({ setFieldValue, lietke, lietkeCustom }) => {
  return (
    <Section>
      <Title title="Phần 2" subtitle="Liệt kê các khó chịu" />
      <SelectGroup
        name={lietkeName}
        value={lietke}
        options={lietkeOptions}
        label={DottedLabel}
        onChange={value => setFieldValue(lietkeName, value)}
        gap="small"
        margin={{ vertical: "small" }}
      />

      {lietke.indexOf("99") !== -1 && (
        <TextInput
          value={lietkeCustom}
          className="p-2"
          onChange={event => {
            setFieldValue(lietkeCustomName, event.target.value);
          }}
        />
      )}
    </Section>
  );
};

const levelOptions = [
  { label: "0", value: "0" },
  { label: "1", value: "1" },
  { label: "2", value: "2" },
  { label: "3", value: "3" },
];

const Part3 = ({ values, selected }) => {
  return (
    <Section>
      <Title title="Phần 3" subtitle="Phiếu ghi nhận chỉ số Child-OIDP" />

      <table className="w-full table-fixed my-4">
        <thead>
          <tr className="h-12 border-b-2">
            <th scope="col" className="font-bold center">
              Các hoạt động
            </th>
            <th scope="col" className="font-bold center">
              Mức trầm trọng
            </th>
            <th scope="col" className="font-bold center">
              Tần suất
            </th>
            <th scope="col" className="font-bold center">
              Nguyên nhân
            </th>
          </tr>
        </thead>
        <tbody>
          {hoatDong.map(row => {
            return (
              <tr
                key={row.value}
                className="border-b border-light-6 hover:bg-light-1"
              >
                <td scope="row" className="py-4">
                  <strong>
                    {row.value}. {row.label}
                  </strong>
                  <br /> {row.secondaryLabel}
                </td>
                <td>
                  <Field
                    name={`${row.value}-mucdo`}
                    render={({ field, form: { setFieldValue } }) => {
                      return (
                        <RadioGroup
                          name={field.name}
                          onChange={value => {
                            setFieldValue(field.name, value);
                          }}
                          options={levelOptions}
                          value={field.value}
                          direction="row"
                          justify="center"
                          flex
                          gap="small"
                        />
                      );
                    }}
                  />
                </td>
                <td>
                  <Field
                    name={`${row.value}-tansuat`}
                    render={({ field, form: { setFieldValue } }) => {
                      const mucdo = values[`${row.value}-mucdo`];
                      if (["1", "2", "3"].indexOf(mucdo) === -1) {
                        return null;
                      }
                      return (
                        <RadioGroup
                          name={field.name}
                          onChange={value => {
                            setFieldValue(field.name, value);
                          }}
                          options={levelOptions}
                          value={field.value}
                          direction="row"
                          justify="center"
                          flex
                          gap="small"
                        />
                      );
                    }}
                  />
                </td>
                <td className="max-w-xs">
                  <Field
                    name={`${row.value}-nguyennhan`}
                    render={({ field, form: { setFieldValue } }) => {
                      const mucdo = values[`${row.value}-mucdo`];

                      if (["1", "2", "3"].indexOf(mucdo) === -1) {
                        return null;
                      }
                      return (
                        <SelectGroup
                          name={field.name}
                          options={selected}
                          value={field.value}
                          onChange={value => setFieldValue(field.name, value)}
                          gap="small"
                          margin={{ vertical: "small" }}
                        />
                      );
                    }}
                  />
                </td>
              </tr>
            );
          })}
        </tbody>
      </table>
    </Section>
  );
};

const findLabelFromId = id => {
  const row = hoatDong.find(row => row.value === id);
  return row.label;
};

const noop = function() {
  return new Promise(resolve => resolve());
};

function isEmpty(obj) {
  for (var key in obj) {
    if (obj.hasOwnProperty(key)) return false;
  }
  return true;
}

const itemIdToName = id => {
  switch (id) {
    case "coKhoChiu":
      return "Có khó chịu hay không - Phần 1";
    case "lietke":
      return "Liệt kê các khó chịu - Phần 2";
    case "lietkeCustom":
      return "Khó chịu khác - Phần 2";
    default:
      const value = hoatDong.find(row => row.value === id.slice(0, 1));

      if (/^[1-8]-mucdo$/.test(id)) {
        return `Mức độ ${id.slice(0, 1)} - ${value && value.label}`;
      } else if (/^[1-8]-nguyennhan$/.test(id)) {
        return `Nguyên nhân ${id.slice(0, 1)} - ${value && value.label}`;
      } else if (/^[1-8]-tansuat$/.test(id)) {
        return `Tần suất ${id.slice(0, 1)} - ${value && value.label}`;
      }
      return id;
  }
};
const FormChildOIDP = ({
  initialValues = blankInitialValues,
  onSave = noop,
}) => (
  <Formik
    initialValues={initialValues}
    onSubmit={(values, { setSubmitting }) => {
      new Promise((resolve, reject) => {
        const validateResult = handleValidation(values);
        if (validateResult.type === "SUCCESS") {
          notify.success("Kiểm tra hoàn tất");
          resolve(values);
        } else if (validateResult.type === "REQUIRED") {
          notify.error("Các mục dưới đây chưa hoàn tất", () => (
            <ul>
              {validateResult.value.map(item => (
                <li key={item} className="py-2">
                  {itemIdToName(item)}
                </li>
              ))}
            </ul>
          ));
          reject(validateResult);
        } else if (validateResult.type === "EXHAUSTIVE_CHECK") {
          console.log(validateResult.value);
          notify.error("Chưa chọn đủ các nguyên nhân khó chịu", () => (
            <div>
              Các khó chịu dưới đây phải nguyên nhân gây khó chịu của ít nhất
              một hoạt động
              <ul>
                {validateResult.value.map(item => (
                  <li key={item} className="py-2">
                    {lietkeOptions.find(row => row.value === item).label}
                  </li>
                ))}
              </ul>
            </div>
          ));
          reject(validateResult);
        } else {
          notify.error("Có lỗi xảy ra", () =>
            JSON.stringify(validateResult, null, 2)
          );
          reject(validateResult);
        }
      })
        .then(values => onSave(values, values.draft))
        .then(() => setSubmitting(false))
        .catch(err => {
          setSubmitting(false);
          console.log(err);
        });
    }}
    validateOnChange={false}
    validateOnBlur={false}
  >
    {({
      values,
      errors,
      touched,
      handleChange,
      handleBlur,
      handleSubmit,
      isSubmitting,
      setFieldValue,
    }) => {
      const selected = values.lietke;
      return (
        <Box>
          <Form>
            <MountPortal id="footerAction">
              <Box justifyContent="end" direction="row" alignItems="center">
                <FormikAutosave
                  values={values}
                  render={({ type }) => {
                    switch (type) {
                      case "INITIAL":
                        return null;
                      case "SAVING":
                        return "Đang lưu";
                      case "SUCCESS":
                        return "Đã lưu";
                      case "ERROR":
                        return "Có lỗi xảy ra khi lưu";
                    }
                  }}
                  onSave={value => onSave(value, true)}
                />
                <Button
                  primary
                  label="Kiểm tra"
                  type="submit"
                  size="small"
                  className="font-bold"
                  onClick={handleSubmit}
                  disabled={isSubmitting}
                  color="white"
                  margin={{ left: "small" }}
                />
              </Box>
            </MountPortal>

            <Box
              direction="row"
              justifyContent="center"
              className="mb-6 lg:mb-12"
            >
              <Heading level={2} textAlign="center" size="medium">
                Bảng câu hỏi về những khó chịu từ răng miệng
              </Heading>
            </Box>

            <Section>
              <Title
                title="Phần 1"
                subtitle="Trong 3 tháng qua các em có thấy khó chịu hay đau ở vùng răng miệng của mình hay không?"
              />
              <FastField
                name="coKhoChiu"
                render={({ field }) => {
                  return (
                    <RadioGroup
                      name={field.name}
                      value={field.value}
                      options={[
                        { label: "Có", value: "1" },
                        { label: "Không", value: "0" },
                      ]}
                      onChange={value => setFieldValue(field.name, value)}
                      direction="row"
                      gap="medium"
                      margin={{ vertical: "small" }}
                    />
                  );
                }}
              />
            </Section>

            {values.coKhoChiu === "1" && (
              <>
                <Part2
                  setFieldValue={setFieldValue}
                  lietke={values.lietke}
                  lietkeCustom={values.lietkeCustom}
                />
                {Boolean(values.lietke.length) && (
                  <Part3
                    values={values}
                    selected={lietkeOptions.filter(
                      ({ value }) => selected.indexOf(value) !== -1
                    )}
                  />
                )}
              </>
            )}
          </Form>
        </Box>
      );
    }}
  </Formik>
);

export default FormChildOIDP;
