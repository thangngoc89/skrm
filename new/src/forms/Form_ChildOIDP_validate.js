export const PART_1_REQUIRED = { type: "PART_1_REQUIRED" };
export const PART_2_REQUIRED = { type: "PART_2_REQUIRED" };
export const REQUIRED_MUCDO = value => ({
  type: "REQUIRED_MUCDO",
  value,
});
export const REQUIRED_TANSUAT = value => ({
  type: "REQUIRED_TANSUAT",
  value,
});
export const REQUIRED_NGUYENNHAN = value => ({
  type: "REQUIRED_NGUYENNHAN",
  value,
});
export const UNSASTIFIED_SUM_NGUYENNHAN = value => ({
  type: "UNSASTIFIED_SUM_NGUYENNHAN",
  value: value,
});
export const SUCCESS = value => ({
  type: "SUCCESS",
  value,
});

export const REQUIRED = {
  type: "REQUIRED",
};

const truthyLevel = ["1", "2", "3"];

const isTruthyLevel = level => {
  return truthyLevel.indexOf(level) !== -1;
};

export const validateHoatdong = ({
  mucdo,
  tansuat,
  nguyennhan,
  keyMucdo,
  keyTansuat,
  keyNguyenNhan,
  lietke,
}) => {
  if (mucdo === "0") {
    return {};
  } else if (mucdo === null) {
    return { [keyMucdo]: REQUIRED };
  } else {
    if (!isTruthyLevel(tansuat)) {
      return { [keyTansuat]: REQUIRED };
    } else if (nguyennhan.length === 0) {
      return { [keyNguyenNhan]: REQUIRED };
    } else {
      /* Filter unselected NGUYENNHAN */
      const nguyennhanFiltered = nguyennhan.filter(
        n => lietke.indexOf(n) !== -1
      );

      if (nguyennhanFiltered.length === 0) {
        return { [keyNguyenNhan]: REQUIRED };
      }
      return {};
    }
  }
};
/*
 * @return
 *   | { type: "PART_1_REQUIRED" }
 *   | { type: "PART_2_REQUIRED" }
 *   | { type: "REQUIRED_MUCDO", value: int}
 *   | { type: "REQUIRED_TANSUAT", value: int}
 *   | { type: "REQUIRED_NGUYENNHAN", value: int}
 *   | { type: "UNSASTIFIED_SUM_NGUYENNHAN", value: array(int) }
 *   | { type: "SUCCESS", value}
 */
export const validate = values => {
  if (values.coKhoChiu !== "1" && values.coKhoChiu !== "0") {
    return { coKhoChiu: REQUIRED };
  } else if (values.coKhoChiu === "0") {
    return {};
  }

  const lietke = values.lietke;

  if (Array.isArray(lietke) && lietke.length === 0) {
    return { lietke: REQUIRED };
  } else if (lietke.indexOf("99") !== -1 && lietke.lietkeCustom === "") {
    return { lietkeCustom: REQUIRED };
  }

  return [1, 2, 3, 4, 5, 6, 7, 8].reduce((acc, i) => {
    const keyMucdo = i + "-mucdo";
    const keyTansuat = i + "-tansuat";
    const keyNguyenNhan = i + "-nguyennhan";

    const valueMucdo = values[keyMucdo];
    const valueTansuat = values[keyTansuat];
    const valueNguyenNhan = values[keyNguyenNhan];

    const result = validateHoatdong({
      mucdo: valueMucdo,
      tansuat: valueTansuat,
      nguyennhan: valueNguyenNhan,
      keyMucdo,
      keyTansuat,
      keyNguyenNhan,
      lietke: values.lietke,
    });
    return { ...acc, ...result };
  }, {});
};
