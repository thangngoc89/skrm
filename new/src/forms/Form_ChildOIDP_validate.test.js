import { validate, REQUIRED } from "./Form_ChildOIDP_validate";
import { blankInitialValues } from "./Form_ChildOIDP";

describe("part 1 is null", () => {
  test("it should require part 1", () => {
    const values = blankInitialValues;
    expect(values.coKhoChiu).toBe(null);
    expect(validate(blankInitialValues)).toEqual({
      coKhoChiu: REQUIRED,
    });
  });
});

describe("part 1 is falsy", () => {
  test("it should be success", () => {
    const values = {
      ...blankInitialValues,
      coKhoChiu: "0",
    };

    expect(validate(values)).toEqual({});
  });
});

describe("part 1 is truthy", () => {
  describe("part 2 is empty", () => {
    test("it should require part 2", () => {
      const values = {
        ...blankInitialValues,
        coKhoChiu: "1",
      };
      expect(validate(values)).toEqual({ lietke: REQUIRED });
    });
  });
  describe("part 2 is selected", () => {
    test("it should not requires part 2", () => {
      const values = {
        ...blankInitialValues,
        coKhoChiu: "1",
        lietke: ["1", "2", "3"],
      };
      expect(validate(values).lietke).toBe(undefined);
    });
  });
});

describe("part 3", () => {
  const defaultMessage = {
    "1-mucdo": REQUIRED,
    "2-mucdo": REQUIRED,
    "3-mucdo": REQUIRED,
    "4-mucdo": REQUIRED,
    "5-mucdo": REQUIRED,
    "6-mucdo": REQUIRED,
    "7-mucdo": REQUIRED,
    "8-mucdo": REQUIRED,
  };
  test("MUCDO = null => it should require MUCDO", () => {
    const values = {
      ...blankInitialValues,
      coKhoChiu: "1",
      lietke: ["1", "2", "3"],
      "1-mucdo": null,
    };
    expect(validate(values)["1-mucdo"]).toEqual(REQUIRED);
    expect(validate(values)).toMatchSnapshot();
  });
  test(`MUCDO = "0" => it should not requires NGUYENNHAN or TANSUAT`, () => {
    const values = {
      ...blankInitialValues,
      coKhoChiu: "1",
      lietke: ["1", "2", "3"],
      "1-mucdo": "0",
    };
    const result = validate(values);
    const allKeysStartWithOne = Object.keys(result).find(key =>
      key.includes("1-")
    );
    expect(allKeysStartWithOne).toBe(undefined);
  });

  test("MUCDO = truthy => it should require TANSUAT", () => {
    const values = {
      ...blankInitialValues,
      coKhoChiu: "1",
      lietke: ["1", "2", "3"],
      "1-mucdo": "1",
    };
    expect(validate(values)["1-mucdo"]).toBe(undefined);
    expect(validate(values)["1-tansuat"]).toEqual(REQUIRED);
    expect(validate(values)).toMatchSnapshot();
  });
  test("MUCDO = truthy, TANSUAT = truthy => it should require NGUYENNHAN", () => {
    const values = {
      ...blankInitialValues,
      coKhoChiu: "1",
      lietke: ["1", "2", "3"],
      "1-mucdo": "1",
      "1-tansuat": "2",
    };
    expect(validate(values)["1-mucdo"]).toBe(undefined);
    expect(validate(values)["1-tansuat"]).toBe(undefined);
    expect(validate(values)["1-nguyennhan"]).toEqual(REQUIRED);
    expect(validate(values)).toMatchSnapshot();
  });
  test("MUCDO = truthy, TANSUAT = truthy, NGUYENNHAN = truthy => it should pass", () => {
    const values = {
      ...blankInitialValues,
      coKhoChiu: "1",
      lietke: ["1", "2", "3"],
      "1-mucdo": "1",
      "1-tansuat": "2",
      "1-nguyennhan": ["1"],
    };
    expect(validate(values)["1-mucdo"]).toBe(undefined);
    expect(validate(values)["1-tansuat"]).toBe(undefined);
    expect(validate(values)["1-nguyennhan"]).toBe(undefined);
    expect(validate(values)).toMatchSnapshot();
  });
  test("it should filter unselected NGUYENHAN", () => {
    const values = {
      ...blankInitialValues,
      coKhoChiu: "1",
      lietke: ["1", "2", "3"],
      "1-mucdo": "1",
      "1-tansuat": "2",
      "1-nguyennhan": ["4"],
    };
    expect(validate(values)["1-nguyennhan"]).toEqual(REQUIRED);
  });

  test("it should do an exhaustive check on NGUYENNHAN", () => {
    const values = {
      coKhoChiu: "1",
      lietke: ["1", "2", "3", "4"],
      lietkeCustom: "",
      "1-mucdo": "0",
      "1-tansuat": null,
      "1-nguyennhan": [],
      "2-mucdo": "0",
      "2-tansuat": null,
      "2-nguyennhan": [],
      "3-mucdo": "0",
      "3-tansuat": null,
      "3-nguyennhan": [],
      "4-mucdo": "0",
      "4-tansuat": null,
      "4-nguyennhan": [],
      "5-mucdo": "0",
      "5-tansuat": null,
      "5-nguyennhan": [],
      "6-mucdo": "0",
      "6-tansuat": null,
      "6-nguyennhan": [],
      "7-mucdo": "1",
      "7-tansuat": "1",
      "7-nguyennhan": ["2", "3"],
      "8-mucdo": "2",
      "8-tansuat": "1",
      "8-nguyennhan": ["1"],
    };
    expect(validate(values)).toEqual("foo");
  });
});
