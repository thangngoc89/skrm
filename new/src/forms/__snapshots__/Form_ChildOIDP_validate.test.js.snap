// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`part 3 MUCDO = null => it should require MUCDO 1`] = `
Object {
  "1-mucdo": Object {
    "type": "REQUIRED",
  },
  "2-mucdo": Object {
    "type": "REQUIRED",
  },
  "3-mucdo": Object {
    "type": "REQUIRED",
  },
  "4-mucdo": Object {
    "type": "REQUIRED",
  },
  "5-mucdo": Object {
    "type": "REQUIRED",
  },
  "6-mucdo": Object {
    "type": "REQUIRED",
  },
  "7-mucdo": Object {
    "type": "REQUIRED",
  },
  "8-mucdo": Object {
    "type": "REQUIRED",
  },
}
`;

exports[`part 3 MUCDO = truthy => it should require TANSUAT 1`] = `
Object {
  "1-tansuat": Object {
    "type": "REQUIRED",
  },
  "2-mucdo": Object {
    "type": "REQUIRED",
  },
  "3-mucdo": Object {
    "type": "REQUIRED",
  },
  "4-mucdo": Object {
    "type": "REQUIRED",
  },
  "5-mucdo": Object {
    "type": "REQUIRED",
  },
  "6-mucdo": Object {
    "type": "REQUIRED",
  },
  "7-mucdo": Object {
    "type": "REQUIRED",
  },
  "8-mucdo": Object {
    "type": "REQUIRED",
  },
}
`;

exports[`part 3 MUCDO = truthy, TANSUAT = truthy => it should require NGUYENNHAN 1`] = `
Object {
  "1-nguyennhan": Object {
    "type": "REQUIRED",
  },
  "2-mucdo": Object {
    "type": "REQUIRED",
  },
  "3-mucdo": Object {
    "type": "REQUIRED",
  },
  "4-mucdo": Object {
    "type": "REQUIRED",
  },
  "5-mucdo": Object {
    "type": "REQUIRED",
  },
  "6-mucdo": Object {
    "type": "REQUIRED",
  },
  "7-mucdo": Object {
    "type": "REQUIRED",
  },
  "8-mucdo": Object {
    "type": "REQUIRED",
  },
}
`;

exports[`part 3 MUCDO = truthy, TANSUAT = truthy, NGUYENNHAN = truthy => it should pass 1`] = `
Object {
  "2-mucdo": Object {
    "type": "REQUIRED",
  },
  "3-mucdo": Object {
    "type": "REQUIRED",
  },
  "4-mucdo": Object {
    "type": "REQUIRED",
  },
  "5-mucdo": Object {
    "type": "REQUIRED",
  },
  "6-mucdo": Object {
    "type": "REQUIRED",
  },
  "7-mucdo": Object {
    "type": "REQUIRED",
  },
  "8-mucdo": Object {
    "type": "REQUIRED",
  },
}
`;
